--- a/hw/pc.c	2009-12-29 21:46:34.000000000 +0100
+++ b/hw/pc.c	2010-01-09 00:01:15.691501649 +0100
@@ -54,9 +54,9 @@
 /* Show multiboot debug output */
 //#define DEBUG_MULTIBOOT
 
-#define BIOS_FILENAME "bios.bin"
-#define EXTBOOT_FILENAME "extboot.bin"
-#define VAPIC_FILENAME "vapic.bin"
+#define BIOS_FILENAME "bios-kvm.bin"
+#define EXTBOOT_FILENAME "extboot-kvm.bin"
+#define VAPIC_FILENAME "vapic-kvm.bin"
 
 #define PC_MAX_BIOS_SIZE (4 * 1024 * 1024)
 
