diff -Naur qemu-kvm-devel-88/Makefile qemu-kvm-devel-88.slb/Makefile
--- a/Makefile	2009-07-12 13:41:41.000000000 +0200
+++ b/Makefile	2009-07-14 17:24:33.000000000 +0200
@@ -165,7 +165,7 @@
 audio-obj-$(CONFIG_FMOD) += fmodaudio.o
 audio-obj-$(CONFIG_ESD) += esdaudio.o
 audio-obj-$(CONFIG_PA) += paaudio.o
-audio-obj-$(AUDIO_PT_INT) += audio_pt_int.o
+audio-obj-y += audio_pt_int.o
 audio-obj-y += wavcapture.o
 obj-y += $(addprefix audio/, $(audio-obj-y))
 
